pkg/kubelet/checkpointmanager/checkpoint_manager.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package checkpointmanager
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager/errors"
+	utilstore "k8s.io/kubernetes/pkg/kubelet/util/store"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// Checkpoint provides the process checkpoint data
+type Checkpoint interface {
+	MarshalCheckpoint() ([]byte, error)
+	UnmarshalCheckpoint(blob []byte) error
+	GetChecksum() uint64
+	UpdateChecksum()
```

Comment:
I think we can include updating the checksum as part of the MarshalCheckpoint function.

Reply:
Reason for keeping it separate is to make checksum as part of common checkpoint-manager framework and keeping it explicitly in the interface to ensure that all implementers are going to implement checksum functionality for the sake of consistency. Same is the case with GetChecksum and VerifyChecksum. These checksum related apis can be removed from interface and then an implementer submodule may or may not use checksums.
 
What you  guys think?

Reply:
WDYT about implementing a seperate `struct ChecksumCheckpoint{data interface{}, checksum uint64}` that implements Checkpoint, and handles the checksumming?  Then each library that wants to implement it can just use a wrapper around ChecksumCheckpoint to convert from interface{} to whatever type they need.

Reply:
I think more generally, checksumming shouldn't be something that is leaked to implementations.  It is an implementation detail of Reading and Writing a checkpoint from disk.

Reply:
true. OTOH, Since each implementation has different checkpoint data, it is not possible for manager to handle checksum without depending(leaking) on implementers.

Reply:
Makes sense.  I didn't realize json marshalling/unmarshalling doesn't work well with interface{}... which means you can't have a gereral checkpointer that does marshalling/unmarshalling and checkpointing for all implementations.

Reply:
I think it is enough to include UpdateChecksum() and VerifyChecksum() in the API, as both of those are used.

Reply:
Take a look at https://github.com/kubernetes/kubernetes/commit/5bde42a9177627999d6cee00c0768228425dde49.  I broke out the checksumming into its own package, and avoid casting to types by adding PodCheckpoint and DeviceManagerCheckpoint interfaces.